File: synteny/psl_merger.py
#!/usr/bin/env python

from argparse import ArgumentParser
from itertools import groupby
import time

from pycbio.hgdata.psl import PslReader
from pycbio.hgdata.psl import Psl

def get_blocks_set(psl_file):
    blocks = []
    for psl in PslReader(psl_file):
        blocks += psl.blocks
    return set(blocks)

#assume a.start < b.start
def are_syntenic(a, b):
    return a.qEnd <= b.qStart and\
            a.tEnd <= b.tStart and\
             a.psl.tName == b.psl.tName and\
                a.psl.strand == b.psl.strand


def is_not_overlapping_ordered_pair(a, b, threshold=5000):
    return are_syntenic(a, b) and\
             0 <= b.qStart - a.qEnd < threshold and\
              0 <=  b.tStart - a.tEnd < threshold


def get_next(pos, query_group, max_anchor_distance=5000):
    f = []
    for i in range(pos+1, len(query_group)):
        if is_not_overlapping_ordered_pair(query_group[pos], query_group[i], max_anchor_distance):
            if not f:
                f.append(i)
            elif f and is_not_overlapping_ordered_pair(query_group[f[0]], query_group[i], max_anchor_distance):
                return f
            else:
                f.append(i)
    return f


'''
dag is a dict that for a given vertex stores all its possible next verties
hidden_vertices is a set of vertices that are already in paths
'''
def weigh_dag(group, dag, hidden_vertices, max_anchor_distance):
    #weight of the edge equals length 
    #of the next block
    #weight of a vertice equals 
    #estimated weight: w_j < w_i + w_e(ij) =>
    #updated w_j
    #also remember how we came here: 
    #(prev_vertex, weight)
    weighted_dag = {}
    for i in range(len(group)):
        if i in hidden_vertices:
            continue
        #print i, group[i], len(group)
        if not i in dag:
            nexts = get_next(i, group, max_anchor_distance)
            dag[i] = nexts
        else:
            nexts = dag[i]
        #if never visited this vertex then 
        #its weight equals to its size
        #because otherwise we will never count its size
        if not i in weighted_dag:
            weighted_dag[i] = (-1, group[i].size)
        for j in nexts:
            if j in hidden_vertices:
                continue
            alternative_weight = weighted_dag[i][1] + group[j].size
            if not j in weighted_dag or weighted_dag[j][1] < alternative_weight: 
                #w_i + weight of the next edge 
                weighted_dag[j] = (i, alternative_weight)
    return weighted_dag

def traceback(weighted_dag, hidden, group):
    #get the heaviest path weight
    start_vertex = max(weighted_dag.items(), key=lambda x:x[1][1])[0]
    path = [start_vertex]
    prev_vertex = weighted_dag[start_vertex][0]
    while prev_vertex != -1:
        path.append(prev_vertex)
        prev_vertex = weighted_dag[prev_vertex][0]
    hidden.update(set(path))
    return map(lambda x: group[x], path)[::-1]

def dag_merge(psl, min_block_breath, max_anchor_distance):
    blocks = get_blocks_set(psl)
    blocks = sorted(blocks, key=lambda x: x.psl.qName)
    blocks_grouped_by_query = map(lambda x:list(x[1]), groupby(blocks, key=lambda x:x.psl.qName))
    paths = []
    for group in blocks_grouped_by_query:
        dag = {}
        hidden_vertices = set()
        group = sorted(group, key=lambda x: x.qStart)
        while len(group) != len(hidden_vertices):
            ts = time.time()
            weighted_dag = weigh_dag(group, dag, hidden_vertices, max_anchor_distance)
            path = traceback(weighted_dag, hidden_vertices, group)
            if not path:
                break
            qLen = path[-1].qEnd - path[0].qStart
            tLen = path[-1].tEnd - path[0].tStart
            if qLen >= min_block_breath and tLen >= min_block_breath:
                paths.append(path)
            #think if we should keep path that is not complies with lengths bounds
            #we could traverse these vertices later in other paths
            #for e in path:
            #    group.remove(e)
    return paths
            
def construct_psl(blocks):
    psl = Psl()
    psl.match = sum(map(lambda x: x.qEnd-x.qStart, blocks))
    psl.misMatch = 0
    psl.repMatch = 0
    psl.nCount = 0
    psl.qNumInsert = len(filter(lambda x: x>0, map(lambda x: x[1].qStart-x[0].qEnd, zip(blocks,blocks[1:]))))
    psl.qBaseInsert = sum(map(lambda x: x[1].qStart-x[0].qEnd, zip(blocks,blocks[1:])))
    psl.tNumInsert = len(filter(lambda x: x>0, map(lambda x: x[1].tStart-x[0].tEnd, zip(blocks,blocks[1:]))))
    psl.tBaseInsert = sum(map(lambda x: x[1].tStart-x[0].tEnd, zip(blocks,blocks[1:])))
    psl.qName = blocks[0].psl.qName
    psl.qSize = blocks[0].psl.qSize
    psl.qStart = blocks[0].qStart
    psl.qEnd = blocks[-1].qEnd
    psl.tName = blocks[0].psl.tName
    psl.tSize = blocks[0].psl.tSize
    psl.strand = blocks[0].psl.strand
    if psl.strand == '++':
        psl.tStart = blocks[0].tStart
        psl.tEnd = blocks[-1].tEnd
    elif psl.strand == '+-':
        psl.tEnd = psl.tSize - blocks[0].tStart
        psl.tStart = psl.tSize - blocks[-1].tEnd
    psl.blockCount = len(blocks)
    for b in blocks:
        b.psl = psl
    psl.blocks = blocks
    return psl
    
if __name__ == '__main__':
    parser = ArgumentParser()
    parser.add_argument('min_block_size', nargs='?', default=5000, type=int)
    parser.add_argument('max_anchor_distance', nargs='?', default=5000, type=int)
    parser.add_argument('psl')
    parser.add_argument('out')
    args = parser.parse_args()
    print 'dag merge...'
    print 'using min_block_size = ', args.min_block_size, \
        'max_anchor_distance =', args.max_anchor_distance
    merged = dag_merge(args.psl, args.min_block_size, args.max_anchor_distance)
    print 'storing output...'
    with open(args.out, 'w') as f:
        for blocks in merged:
           psl = construct_psl(blocks)
           f.write('\t'.join(psl.toRow())+'\n')

File: c++/psl_merger.cpp
#include <iostream>
#include <set>
#include <map>
#include <vector>
#include <utility>
#include <algorithm>
#include <ctime> 

#include "psl_reader.h"


//assume a.start < b.start
bool are_syntenic(const PslBlock& a, const PslBlock& b){
    return a.qEnd <= b.qStart &&
            a.tEnd <= b.tStart &&
             a.tName == b.tName &&
                a.strand == b.strand;
}

bool is_not_overlapping_ordered_pair(const PslBlock& a, const PslBlock& b, const int threshold=5000) {
    //std::cout << a.qStart << " " << b.qStart << " " << are_syntenic(a, b) << " " << a.tStart << " " << b.tStart << " "
    //    << (0 <= b.qStart - a.qEnd < threshold) << " " << (0 <=  b.tStart - a.tEnd < threshold) << " " << b.tStart - a.tEnd << " " << threshold  << std::endl;
    return are_syntenic(a, b) &&
             0 <= b.qStart - a.qEnd && 
                b.qStart - a.qEnd < threshold && 
                 0 <= b.tStart - a.tEnd &&
                    b.tStart - a.tEnd < threshold;
}

std::vector<int> get_next(const int pos, const std::vector<PslBlock>& queryGroup,
                 const int maxAnchorDistance=5000) {
    std::vector<int> f;
    //for (auto l : queryGroup) std::cout << l.qStart << " ";
    //std::cout << std::endl;
    for (auto i = pos + 1; i < queryGroup.size(); ++i) {
        if (is_not_overlapping_ordered_pair(queryGroup[pos], queryGroup[i], maxAnchorDistance)) {
            if (f.empty()) f.push_back(i);
            else {
                if (not f.empty() && is_not_overlapping_ordered_pair(queryGroup[f[0]], queryGroup[i], maxAnchorDistance)) {
                    //!!!!!!!!
                    ///for (auto l : f) std::cout << l << " (" << queryGroup[l].qStart << "," << queryGroup[l].qEnd << "," 
                    //                        << queryGroup[l].tStart << "," <<  queryGroup[l].tEnd << "," << queryGroup[l].strand << ") ";
                    //std::cout << std::endl;
                    return f;
                }
                else {
                    f.push_back(i);
                }
            }
        }
    }
    
    return f;
}

/*
dag is a dict that for a given vertex stores all its possible next verties
hidden_vertices is a set of vertices that are already in paths
weight of the edge equals length
of the next block
weight of a vertice equals
estimated weight: w_j < w_i + w_e(ij) =>
updated w_j
also remember how we came here:
(prev_vertex, weight)
*/
std::map<int, std::pair<int, int> > weigh_dag(const std::vector<PslBlock>& group, 
                                             std::map<int, std::vector<int> >& dag, 
                                             const std::set<int>& hiddenVertices,
                                             const int maxAnchorDistance){
    std::map<int,std::pair<int,int> > weightedDag;
    for (int i = 0; i < group.size(); ++i) {
        if (hiddenVertices.count(i)) continue; 
        std::vector<int> nexts;
        if (not dag.count(i)) {
            nexts = get_next(i, group, maxAnchorDistance);
            dag[i] = nexts;
        }
        else {
            nexts = dag[i];
       }
       //std::cout << i;
       //for (auto n : nexts) {
       //std::cout << " " << n;
       //}
       //std::cout << std::endl;
       /*
       if never visited this vertex then
       its weight equals to its size
       because otherwise we will never count its size
       */
       if (not weightedDag.count(i))
            weightedDag[i] = std::make_pair(-1, group[i].size);
       for (auto j: nexts){
            if (hiddenVertices.count(j)) continue;
            auto alternativeWeight = weightedDag[i].second + group[j].size;
            if (not weightedDag.count(j) or weightedDag[j].second < alternativeWeight)
                //w_i + weight of the next edge 
                weightedDag[j] = std::make_pair(i, alternativeWeight);
       }
    }
    //std::cout << "weighted dag\n";
    //for (auto it = weightedDag.cbegin(); it != weightedDag.cend(); ++it) {
    //    std::cout << it ->first << " " << group[it->first].qStart <<  " " << group[it->first].qEnd << " " << group[it->first].tStart 
    //        << " "<< it->second.first << " " << it->second.second << std::endl;
    //}
    return weightedDag; 
}

int get_maxed_vertex(const std::map<int, std::pair<int, int> >& weightedDag) {
    int max = weightedDag.cbegin()->second.second;
    int maxPos = weightedDag.cbegin()->first;
    for (auto it = weightedDag.cbegin(); it != weightedDag.cend(); ++it) {
        //std::cout << "it->second.second " << it->second.second << std::endl;
        if (it->second.second >= max) {
            max = it->second.second;
            maxPos = it->first;
        }
    }
    //std::cout << "max pos " << maxPos << std::endl;
    return maxPos;
}

std::vector<PslBlock> traceback(std::map<int, std::pair<int, int> >& weightedDag, 
                                std::set<int>& hiddenVertices, 
                                const std::vector<PslBlock>& group) {
    //get the heaviest path weight
    auto startVertex = get_maxed_vertex(weightedDag);
    //std::cout << 11 << " " << startVertex << std::endl;
    std::vector<int> path = {startVertex};
    auto prevVertex = weightedDag[startVertex].first;
    //std::cout << 12 << std::endl;
    while (prevVertex != -1) {
        path.push_back(prevVertex);
        prevVertex = weightedDag[prevVertex].first;
    }
    hiddenVertices.insert(path.begin(), path.end());
    std::vector<PslBlock> pslBlockPath;
    //std::cout << "path size " << path.size() << std::endl;
    //for (auto p: path) {
    //    std::cout << group[p].qStart << " " << group[p].qEnd << " " << group[p].strand << " " << group[p].tStart << " " << group[p].tEnd << std::endl;
    //}
    //std::cout << std::endl;
    //std::cout << 4 << std::endl;
    std::transform(path.begin(), path.end(), std::back_inserter(pslBlockPath), [group] (int x) -> PslBlock {return group[x];});
    //std::cout << 5 << std::endl;
    std::reverse(pslBlockPath.begin(),pslBlockPath.end());
    //std::cout << 6 << std::endl;
    return pslBlockPath;
}
struct {    
    bool operator()(PslBlock a, PslBlock b) const
    {
        //return a.qStart < b.qStart && a.tStart < b.tStart;
        if (a.qStart < b.qStart )
            return true;
        else if (a.qStart == b.qStart) {
            return a.tStart <= b.tStart;
        }
        return false;
    }
} qStartLess;

/*struct {    
    bool operator()(PslBlock a, PslBlock b) const
    {
        //return a.qStart < b.qStart && a.tStart < b.tStart;
        return a.tStart < b.tStart;
    }
} tStartLess;
*/

std::vector<std::vector<PslBlock> >  dag_merge(const std::vector<PslBlock>& blocks, const int minBlockBreath, const int maxAnchorDistance){
    std::map<std::string, std::vector<PslBlock> > blocksByQName;
    for (auto block : blocks) blocksByQName[block.qName].push_back(block);
    //for (auto pairs : blocksByQName) {
    //    std::cout << pairs.first << " " << pairs.second.size() << std::endl;
    //}
    std::vector<std::vector<PslBlock> > paths;
    for (auto pairs : blocksByQName) {
        //if (pairs.first != "scaffold1")
        //    continue;
        std::vector<PslBlock> group = pairs.second; 
        std::map<int, std::vector<int> > dag;
        std::set<int> hiddenVertices;
        std::sort(group.begin(), group.end(), qStartLess); 
        //std::stable_sort(group.begin(), group.end(), tStartLess); 
        //for (auto g: group){
        //    std::cout << g.qStart << " " << g.qName << " " << g.tStart << " " << g.tName<< std::endl;
        //}
        //std::cout << std::endl;
        while (hiddenVertices.size() != group.size()){
            //std::cout << "size " << hiddenVertices.size() << std::endl;
            //auto t = clock();
            auto weightedDag = weigh_dag(group, dag, hiddenVertices, maxAnchorDistance);
            //std::cout << "get_next time " << float(clock() - t)/CLOCKS_PER_SEC << std::endl;
            //t = clock();
            auto path = traceback(weightedDag, hiddenVertices, group);
            //std::cout << "traceback time " << float(clock() - t)/CLOCKS_PER_SEC << std::endl;
            //std::cout << path.size() << std::endl;
            if (path.empty()) {
                break;
            }
            auto qLen = path.back().qEnd - path[0].qStart;
            auto tLen = path.back().tEnd - path[0].tStart;
               // std::cout << path.size() << " " << tLen << " " << qLen << " " << minBlockBreath;
            if (qLen >= minBlockBreath && tLen >= minBlockBreath){
                paths.push_back(path);
                //std::cout << " OK" << std::endl;
            }
            //else
            //    std::cout << std::endl;
            //std::cout << pairs.first << " group size " << group.size() << " hidden vertices size " << hiddenVertices.size() << std::endl;
        }
    }
    return paths;
}


int main() {
    //auto path = "/Users/admin/projects/breaks/data/AcinonyxJubatus.FelisCatus.19.psl";
    //auto path = "/hive/groups/recon/projs/felidae_comp/synteny-play-ground/data/felidae/cheetah/AcinonyxJubatus.FelisCatus.19.psl";
    auto path = "/mnt/storagesm/projects/12cats/ksenia/felidae_comp/synteny-play-ground/data/felidae/cheetah/AcinonyxJubatus.FelisCatus.19.psl";
    auto min_block_size = 5000;
    auto max_anchor_distance = 5000;
    /*
    auto blocks = psl_io::get_blocks_set(path);
    std::cout << blocks.size() << "\n";
    for (auto b: blocks) {
        std::cout << b.qStart << " " << b.qEnd << " " << b.tStart << " " << b.tEnd << " " << b.size << "\n";
    }*/
    auto blocks = psl_io::get_blocks_set(path);
    auto merged_blocks = dag_merge(blocks, min_block_size, max_anchor_distance);
    for (auto path : merged_blocks){
        auto psl = psl_io::construct_psl(path);
        std::cout << psl << std::endl;
    } 
    return 0;
}
